internal/recipient/schedule.go

```diff
@@ -3,50 +3,69 @@ package recipient
+	"go.uber.org/zap/zapcore"
 	"time"
 )
 
 type Schedule struct {
-	ID         int64  `db:"id"`
-	Name       string `db:"name"`
-	Members    []*Member
-	MemberRows []*ScheduleMemberRow
+	ID   int64  `db:"id"`
+	Name string `db:"name"`
+
+	Rotations        []*Rotation
+	rotationResolver rotationResolver
 }
 
-type Member struct {
-	TimePeriod   *timeperiod.TimePeriod
-	Contact      *Contact
-	ContactGroup *Group
+// RefreshRotations updates the internally cached rotations.
+//
+// This must be called after the Rotations member was updated for the change to become active.
+func (s *Schedule) RefreshRotations() {
+	s.rotationResolver.update(s.Rotations)
 }
 
-type ScheduleMemberRow struct {
-	ScheduleID   int64         `db:"schedule_id"`
-	TimePeriodID int64         `db:"timeperiod_id"`
-	ContactID    sql.NullInt64 `db:"contact_id"`
-	GroupID      sql.NullInt64 `db:"contactgroup_id"`
+type Rotation struct {
+	ID            int64           `db:"id"`
+	ScheduleID    int64           `db:"schedule_id"`
+	ActualHandoff types.UnixMilli `db:"actual_handoff"`
+	Priority      int32           `db:"priority"`
+	Name          string          `db:"name"`
+	Members       []*RotationMember
 }
 
-func (s *ScheduleMemberRow) TableName() string {
-	return "schedule_member"
+// MarshalLogObject implements the zapcore.ObjectMarshaler interface.
+func (r *Rotation) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
+	encoder.AddInt64("id", r.ID)
+	encoder.AddInt64("schedule_id", r.ScheduleID)
+	encoder.AddInt32("priority", r.Priority)
+	encoder.AddString("name", r.Name)
+	return nil
 }
 
-// GetContactsAt returns the contacts that are active in the schedule at the given time.
-func (s *Schedule) GetContactsAt(t time.Time) []*Contact {
-	var contacts []*Contact
-
-	for _, m := range s.Members {
-		if m.TimePeriod.Contains(t) {
-			if m.Contact != nil {
-				contacts = append(contacts, m.Contact)
-			}
-
-			if m.ContactGroup != nil {
-				contacts = append(contacts, m.ContactGroup.Members...)
-			}
-		}
+type RotationMember struct {
+	ID                int64         `db:"id"`
+	RotationID        int64         `db:"rotation_id"`
+	Contact           *Contact      `db:"-"`
+	ContactID         sql.NullInt64 `db:"contact_id"`
+	ContactGroup      *Group        `db:"-"`
+	ContactGroupID    sql.NullInt64 `db:"contactgroup_id"`
+	TimePeriodEntries map[int64]*timeperiod.Entry
+}
+
+func (r *RotationMember) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
+	encoder.AddInt64("id", r.ID)
+	encoder.AddInt64("rotation_id", r.RotationID)
+	if r.ContactID.Valid {
+		encoder.AddInt64("contact_id", r.ContactID.Int64)
 	}
+	if r.ContactGroupID.Valid {
+		encoder.AddInt64("contact_group_id", r.ContactGroupID.Int64)
+	}
+	return nil
+}
 
-	return contacts
+// GetContactsAt returns the contacts that are active in the schedule at the given time.
+func (s *Schedule) GetContactsAt(t time.Time) []*Contact {
+	return s.rotationResolver.getContactsAt(t)
```

Comment:
For debugging reasons, it would be nice to have a logging message when no `Contacts` are being returned. This might also happen somewhere up the call stack.